python/dgl/frame.py

```diff
@@ -548,7 +559,17 @@ def record_stream(self, stream):
         """
         if F.get_preferred_backend() != "pytorch":
             raise DGLError("record_stream only supports the PyTorch backend.")
-        self.data.record_stream(stream)
+        # Only record streams for materialized tensors otherwise just record for the index
+        if self.index is None:
+            self.data.record_stream(stream)
+        else:
+            if (
+                isinstance(self.index, _LazyIndex)
+                or F.context(self.index) != F.cpu()
+            ):
+                self.index.record_stream(stream)
+            if F.context(self.storage) != F.cpu():
+                self.storage.record_stream(stream)
```

Comment:
If I store the main graph in GPU memory, and then generate subgraphs on a new stream, I wouldn't want to record the stream on the feature storage of the original graph.

I'm not sure of the right course of action however, as adding a deferred stream would seem to just make matters worse.

Reply:
I feel that is the only solution if the column will be materialized later.
I don't think there is an issue of stream recording the main graph.

Reply:
You're right, I was misunderstanding the semantics of record_stream. So for correctness we want to call record stream on both the index and the storage (the only actual tensors associated with a column), but never on `.data`?

Reply:
`.data` is a property (function call) that would trigger the costly `tensor.select` operation. So yes we want to call record stream on index and storage that will be used later if the column gets materialized.

Reply:
This looks good to me. Thanks!